src/System.Private.CoreLib/shared/System/Text/Rune.cs

```diff
@@ -0,0 +1,724 @@
+
+        public static bool operator <=(Rune left, Rune right) => (left.ValueUnsigned <= right.ValueUnsigned);
+
+        public static bool operator >(Rune left, Rune right) => (left.ValueUnsigned > right.ValueUnsigned);
+
+        public static bool operator >=(Rune left, Rune right) => (left.ValueUnsigned >= right.ValueUnsigned);
+
+        // Operators below are explicit because they may throw.
+
+        public static explicit operator Rune(char ch) => new Rune(ch);
+
+        [CLSCompliant(false)]
+        public static explicit operator Rune(uint value) => new Rune(value);
+
+        public static explicit operator Rune(int value) => new Rune(value);
+
+        // Displayed as "'<char>' (U+XXXX)"; e.g., "'e' (U+0065)"
+        private string DebuggerDisplay => FormattableString.Invariant($"'{(IsValid(ValueUnsigned) ? ToString() : "\uFFFD")}' (U+{ValueUnsigned:X4})");
+
+        /// <summary>
+        /// Returns true if and only if this scalar value is ASCII ([ U+0000..U+007F ])
+        /// and therefore representable by a single UTF-8 code unit.
+        /// </summary>
+        public bool IsAscii => UnicodeUtility.IsAsciiCodePoint(ValueUnsigned);
+
+        /// <summary>
+        /// Returns true if and only if this scalar value is within the BMP ([ U+0000..U+FFFF ])
+        /// and therefore representable by a single UTF-16 code unit.
+        /// </summary>
+        public bool IsBmp => UnicodeUtility.IsBmpCodePoint(ValueUnsigned);
+
+        /// <summary>
+        /// Returns the Unicode plane (0 to 16, inclusive) which contains this scalar.
+        /// </summary>
+        public int Plane => UnicodeUtility.GetPlane(ValueUnsigned);
+
+        /// <summary>
+        /// A <see cref="Rune"/> instance that represents the Unicode replacement character U+FFFD.
+        /// </summary>
+        public static Rune ReplacementChar => UnsafeCreate(UnicodeUtility.ReplacementChar);
+
+        /// <summary>
+        /// Returns the length in code units (<see cref="Char"/>) of the
+        /// UTF-16 sequence required to represent this scalar value.
+        /// </summary>
+        /// <remarks>
+        /// The return value will be 1 or 2.
+        /// </remarks>
+        public int Utf16SequenceLength => UnicodeUtility.GetUtf16SequenceLength(ValueUnsigned);
+
+        /// <summary>
+        /// Returns the length in code units (<see cref="Utf8Char"/>) of the
+        /// UTF-8 sequence required to represent this scalar value.
+        /// </summary>
+        /// <remarks>
+        /// The return value will be 1 through 4, inclusive.
+        /// </remarks>
+        public int Utf8SequenceLength => UnicodeUtility.GetUtf8SequenceLength(ValueUnsigned);
+
+        /// <summary>
+        /// Returns the Unicode scalar value as an integer.
+        /// </summary>
+        public int Value => (int)ValueUnsigned;
+
+        internal uint ValueUnsigned { get; }
```

Comment:
The state for types like this should be explicit at the top of the file, not buried in the middle as a getter property. 

I would make this readonly field at the top.

Reply:
Also, autoproperties turn simple field fetches into calls that create busy work for the JIT and may not always get fully optimized out.

Reply:
https://github.com/dotnet/roslyn/issues/15644 😦